cloud/aws/s3inventory/parquet_reader.go

```diff
@@ -2,110 +2,115 @@ package s3inventory
 
-func (p *ParquetInventoryFileReader) Read(n int) ([]InventoryObject, error) {
-	res := make([]InventoryObject, n)
-	actualFields := p.getActualFields()
-	var wg sync.WaitGroup
-	errChan := make(chan error, len(actualFields))
-	for fieldName, path := range actualFields {
-		wg.Add(1)
-		path := path
-		fieldName := fieldName
-		go func() {
-			defer wg.Done()
-			columnRes, _, _, err := p.ReadColumnByPath(path, int64(n))
-			if err != nil {
-				errChan <- fmt.Errorf("failed to read parquet column %s: %w", fieldName, err)
-				return
+func (p *ParquetInventoryFileReader) Read(n int) ([]*InventoryObject, error) {
+	num := mathutil.MinInt64(int64(n), p.GetNumRows()-p.nextRow)
+	p.nextRow += num
+	res := make([]*InventoryObject, num)
+	for fieldName, path := range p.fieldToParquetPath {
+		columnRes, _, dls, err := p.ReadColumnByPath(path, num)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read parquet column %s: %w", fieldName, err)
+		}
+		for i, v := range columnRes {
+			if !isRequired(fieldName) && dls[i] == 0 {
+				// got no value for non-required field, move on
+				continue
 			}
-			for i, v := range columnRes {
-				err := set(&res[i], fieldName, v)
-				if err != nil {
-					errChan <- fmt.Errorf("failed to read parquet column %s: %w", fieldName, err)
-					return
-				}
+			if res[i] == nil {
+				res[i] = NewInventoryObject()
 			}
-		}()
-	}
-	wg.Wait()
-	select {
-	case err := <-errChan:
-		return nil, err
-	default:
+			err := set(res[i], fieldName, v)
+			if err != nil {
+				return nil, fmt.Errorf("failed to read parquet column %s: %w", fieldName, err)
+			}
+		}
 	}
-	close(errChan)
 	return res, nil
 }
 
 func set(o *InventoryObject, f string, v interface{}) error {
 	var err error
 	switch f {
-	case "bucket":
-		var bucket string
-		bucket, err = cast.ToStringE(v)
-		o.Bucket = bucket
-	case "key":
-		var key string
-		key, err = cast.ToStringE(v)
-		o.Key = key
-	case "is_latest":
-		var isLatest bool
-		isLatest, err = cast.ToBoolE(v)
-		o.IsLatest = swag.Bool(isLatest)
-	case "is_delete_marker":
-		var isDeleteMarker bool
-		isDeleteMarker, err = cast.ToBoolE(v)
-		o.IsDeleteMarker = swag.Bool(isDeleteMarker)
-	case "size":
-		var size int64
-		size, err = cast.ToInt64E(v)
-		o.Size = swag.Int64(size)
-	case "last_modified_date":
+	case bucketFieldName:
+		o.Bucket, err = cast.ToStringE(v)
+	case keyFieldName:
+		o.Key, err = cast.ToStringE(v)
+	case isLatestFieldName:
+		o.IsLatest, err = cast.ToBoolE(v)
+	case isDeleteMarkerFieldName:
+		o.IsDeleteMarker, err = cast.ToBoolE(v)
+	case sizeFieldName:
+		o.Size, err = cast.ToInt64E(v)
+	case lastModifiedDateFieldName:
 		var lastModifiedMillis int64
 		lastModifiedMillis, err = cast.ToInt64E(v)
-		o.LastModifiedMillis = swag.Int64(lastModifiedMillis)
-	case "e_tag":
-		var checksum string
-		checksum, err = cast.ToStringE(v)
-		o.Checksum = swag.String(checksum)
+		o.LastModified = swag.Time(time.Unix(lastModifiedMillis/int64(time.Second/time.Millisecond), 0))
```

Comment:
Do we care about losing the subsecond resolution?
```suggestion
		o.LastModified = swag.Time(time.Unix(lastModifiedMillis/int64(1000), (lastModifiedMillis % 1000) * int64(1_000_000))
```

(also you can see that I'm not a fan of pretending that the number of milliseconds per second may change, and I'm trying to sneak that change by you...).

Reply:
Our linter would never allow it !